### ISSUE TITLE
Can `add_unreachable_coverage()` in `mapgen.rs` be simplified?

### ISSUE THREAD

**Issue by richkadel** [rust-lang/rust#79622]
See https://github.com/rust-lang/rust/pull/79109#discussion_r533809308

The PR introduced a solution to adding `Zero` coverage for functions that were processed for MIR, but not codegenned, because they were optimized out and never reached (never called).

The algorithm for finding and injecting the additional code regions, in the right place, is a bit complex.

Can this solution be simplified and/or improved? Are there other sources of compiler state data that might help (or better sources that could be exposed via `cx`, `tcx`, or something else, but aren't yet)?

(Related to Tracking Issue #79121)

**Comment by richkadel** [comment]
cc: @tmandry

**Comment by tmandry** [comment]
It's worth noting that a weakness of the current approach is that crates which only define generic functions won't get zero counters for their unused functions. It would be nice to fix that edge case somehow.

**Comment by richkadel** [comment]
I agree (assuming this is not particularly rare). As we discussed, maybe a way to do that is to forcibly inject a `root` `MonoItem` for an empty, unused, non-generic function, into any crate that doesn't have at least one codegenned non-generic function.

I'm not sure where/how to do that most efficiently, but it's probably not impossible.

@wesleywiser - I think I saw you've worked on code related to `collect_and_partition_mono_items`. I don't know if you have some thoughts on this?
